nixos/modules/services/misc/geoip-databases.nix

```diff
@@ -0,0 +1,179 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.geoip-databases;
+
+  dbBaseUrl = "https://geolite.maxmind.com/download/geoip/database";
+
+  # Use writeScriptBin instead of writeScript, so that argv[0] (logged to the
+  # journal) doesn't include the long nix store path hash. (Prefixing the
+  # ExecStart= command with '@' doesn't work because we start a shell (new
+  # process) that creates a new argv[0].)
+  geoip-updater = pkgs.writeScriptBin "geoip-updater" ''
+    #!${pkgs.stdenv.shell}
+    info()
+    {
+        echo "<6>$@"
+    }
+    error()
+    {
+        echo "<3>$@"
+    }
+    waitNetworkOnline()
+    {
+        ret=1
+        for i in $(seq 6); do
+            info "Checking network connectivity, try $i"
+            curl_out=$("${pkgs.curl.bin}/bin/curl" --silent --show-error "${dbBaseUrl}" 2>&1)
+            if [ $? -eq 0 ]; then
+                ret=0
+                break
+            else
+                info "$curl_out"
+                sleep 10
+            fi
+        done
+        return $ret
+    }
+    fetchDb()
+    {
+        url=$1
+        info "Fetching $url"
+        curl_out=$(cd "${cfg.databaseDir}" &&
+            "${pkgs.curl.bin}/bin/curl" --silent --show-error -LO "$url" 2>&1)
+        if [ $? -ne 0 ]; then
+            error "$curl_out"
+            return 1
+        fi
+    }
+    fetchDatabases()
+    {
+        ret=0
+        for db in ${lib.concatStringsSep " " cfg.databases}; do
+            fetchDb "${dbBaseUrl}/$db" || ret=1
+        done
+        return $ret
+    }
+    unpackDatabases()
+    {
+        (ret=0; cd "${cfg.databaseDir}" &&
+            for db in ${lib.concatStringsSep " " cfg.databases}; do
+                cf=$(basename "$db")
+                case "$cf" in
+                    *.gz)
+                        cmd_out=$("${pkgs.gzip}/bin/gzip" --decompress --force "$cf" 2>&1)
```

Comment:
Does the decompressor replace the databases in-place? If yes, they should probably unpack to a temporary location first and the result is only moved back on success to avoid broken geoip databases.